host/retry_policy_workflow_test.go

```diff
@@ -0,0 +1,319 @@
+	s.NotNil(expirationTime)
+	delta := time.Unix(0, *expirationTime).Sub(now)
+	s.True(delta > 5*time.Second)
+	s.True(delta < 6*time.Second)
+
+	wfExecution, err := s.engine.DescribeWorkflowExecution(createContext(), &types.DescribeWorkflowExecutionRequest{
+		Domain: s.domainName,
+		Execution: &types.WorkflowExecution{
+			WorkflowID: id,
+		},
+	})
+	s.NotNil(wfExecution)
+	delta = time.Unix(0, *wfExecution.WorkflowExecutionInfo.CloseTime).Sub(time.Unix(0, *wfExecution.WorkflowExecutionInfo.StartTime))
+	s.True(delta > 9*time.Second)
+	s.True(delta < 11*time.Second)
+}
+
+func (s *IntegrationSuite) TestWorkflowRetryPolicyContinueAsNewAsRetry() {
+	id := "integration-workflow-retry-policy-continue-as-new-retry-test"
+	wt := "integration-workflow-retry-policy-continue-as-new-retry-test-type"
+	tl := "integration-workflow-retry-policy-continue-as-new-retry-test-tasklist"
+	identity := "worker1"
+
+	workflowType := &types.WorkflowType{}
+	workflowType.Name = wt
+
+	taskList := &types.TaskList{}
+	taskList.Name = tl
+
+	retryPolicy := &types.RetryPolicy{
+		InitialIntervalInSeconds:    1,
+		BackoffCoefficient:          2.0,
+		MaximumIntervalInSeconds:    5,
+		ExpirationIntervalInSeconds: 5,
+		MaximumAttempts:             0,
+		NonRetriableErrorReasons:    []string{"bad-error"},
+	}
+	request := &types.StartWorkflowExecutionRequest{
+		RequestID:                           uuid.New(),
+		Domain:                              s.domainName,
+		WorkflowID:                          id,
+		WorkflowType:                        workflowType,
+		TaskList:                            taskList,
+		Input:                               nil,
+		ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(10),
+		TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(10),
+		Identity:                            identity,
+		RetryPolicy:                         retryPolicy,
+	}
+
+	we, err0 := s.engine.StartWorkflowExecution(createContext(), request)
+	s.Nil(err0)
+
+	s.Logger.Info("StartWorkflowExecution", tag.WorkflowRunID(we.RunID))
+	dtHandler := func(execution *types.WorkflowExecution, wt *types.WorkflowType,
+		previousStartedEventID, startedEventID int64, history *types.History) ([]byte, []*types.Decision, error) {
+		buf := new(bytes.Buffer)
+		return []byte(strconv.Itoa(0)), []*types.Decision{{
+			DecisionType: types.DecisionTypeContinueAsNewWorkflowExecution.Ptr(),
+			ContinueAsNewWorkflowExecutionDecisionAttributes: &types.ContinueAsNewWorkflowExecutionDecisionAttributes{
+				WorkflowType:                        workflowType,
+				TaskList:                            &types.TaskList{Name: tl},
+				Input:                               buf.Bytes(),
+				ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(10),
+				TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(10),
+				RetryPolicy:                         retryPolicy,
+				Initiator:                           types.ContinueAsNewInitiatorRetryPolicy.Ptr(),
```

Comment:
Continue as new is an implementation detail inside cadence server and `Initiator` should not be specified by caller. The decision task handler only needs to return a fail workflow decision and the server will convert it to a continue as new event with the right initiator.